src/org/javarosa/core/model/actions/ActionController.java

```diff
@@ -0,0 +1,87 @@
+package org.javarosa.core.model.actions;
+
+
+import org.javarosa.core.model.FormDef;
+import org.javarosa.core.model.instance.TreeReference;
+import org.javarosa.core.util.externalizable.DeserializationException;
+import org.javarosa.core.util.externalizable.ExtUtil;
+import org.javarosa.core.util.externalizable.ExtWrapListPoly;
+import org.javarosa.core.util.externalizable.ExtWrapMap;
+import org.javarosa.core.util.externalizable.Externalizable;
+import org.javarosa.core.util.externalizable.PrototypeFactory;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+
+/**
+ * Registers actions that should be triggered by certain events, and handles the triggering
+ * of those actions when an event occurs.
+ *
+ * @author Aliza Stone
+ */
+public class ActionController implements Externalizable {
+
+    // map from an event to the actions it should trigger
+    private HashMap<String, List<Action>> eventListeners;
+
+    public ActionController() {
+        this.eventListeners = new HashMap<String, List<Action>>();
```

Comment:
I’m thinking maybe you want to see if the general approach is the right one, before improving the code.
`eventListeners = new HashMap<>();`
So I’ll stop with these.

Reply:
Yes... so any high-level thoughts? 😄